Library/Homebrew/software_spec.rb

```diff
@@ -12,19 +12,20 @@
 require "patch"
 require "compilers"
 require "os/mac/version"
-require "extend/on_os"
+require "extend/on_system"
 
 class SoftwareSpec
   extend T::Sig
 
   extend Forwardable
-  include OnOS
 
   PREDEFINED_OPTIONS = {
     universal: Option.new("universal", "Build a universal binary"),
     cxx11:     Option.new("c++11",     "Build using C++11 mode"),
   }.freeze
 
+  OnSystem.setup_methods! onto: self
```

Comment:
I can't seem to figure out a way to infer it (at least not without adding extra gems or doing a bunch of work that doesn't seem worth it), but if you can figure it out that would definitely be preferable.

Reply:
This might be a stupid question but would it be possible to use the `self.included(base)` callback to setup methods instead? That would mean that you'd only need to `include OnSystem` everywhere it's needed. Just an idea.

Reply:
@Rylan12 Would love to see if @apainintheneck's method works here?

Reply:
So I looked into this last night and this morning, and I'm not totally sure what the right answer is. If we use `self.included(base)`, everything seems to work but I can't figure out how to conditionally define `on_macos` and `on_linux` blocks so we don't get those in the cask DSL where they make no sense. That's probably not a huge deal, but I don't like the idea of having these methods defined when we know we don't need them.

---

I found another option [here](https://medium.com/@eric.programmer/arguments-for-included-modules-in-ruby-8056b9fa2743) which we can make work, but it makes the code a bit less clear and you have to do a weird `include OnSystem.new` thing. Here's how the code for `OnSystem` definition would look:

```ruby
class OnSystem < Module
  def initialize(include_linux: true)
    @include_linux = include_linux
    super()
  end

  def included(base)
    # Existing `setup_methods!` contents except with `base` instead of `onto`
  end
end
```

And then when we include it:

```ruby
# With `on_linux` and `on_macos` methods
include OnSystem.new
# Without `on_linux` and `on_macos` methods
include OnSystem.new include_linux: false
```

In theory, we could also create helper class methods to make this look a bit better:

```ruby
class OnSystem < Module
  def self.with_linux_methods
    new
  end
  def self.without_linux_methods
    new include_linux: false
  end
end

# With `on_linux` and `on_macos` methods
include OnSystem.with_linux_methods
# Without `on_linux` and `on_macos` methods
include OnSystem.without_linux_methods
```

Any thoughts on which of the above options (or keeping as-is) is preferable?

Reply:
I like either of these!

Reply:
Great! I think I'll go with the first for now to get this merged (once I address https://github.com/Homebrew/brew/pull/13451#discussion_r909562616) but I may try to get the second option working in a separate PR

Reply:
Couldn't you just check if the class being passed into the included callback is the Cask DSL class before monkey patching or is that not possible for some reason?

Reply:
Yeah, I think you could. To me, it just doesn't feel like that's where it should be handled. Instead, it seems like the module that includes `OnSystem` should be the one to determine which methods are added. That way, we (ideally) wouldn't have to change the `OnSystem` definition in order to include it in new parts of Homebrew.

But it might be worth doing, anyway since it's better than nothing

Reply:
I can see why maybe you'd want to be more explicit about that instead of hiding that logic inside `OnSystem`.